lightbluetent/admin.py

```diff
@@ -1,39 +1,194 @@
+            "website": society.website,
+            "description": society.description,
+            "welcome_text": society.welcome_text,
+            "logo": society.logo,
+            "banner_text": society.banner_text,
+            "banner_color": society.banner_color,
+            "mute_on_start": society.mute_on_start,
+            "disable_private_chat": society.disable_private_chat
+        }
+
+        return render_template("admin/admin.html", page_title=f"Administration page for { society.short_name }",
+                               short_name=short_name, errors={}, admins=society_admins, **values)
+
+@auth_decorator
+@bp.route("/<short_name>/add", methods=("GET", "POST"))
+def add(short_name):
+    current_crsid = auth_decorator.principal
+    current_user = User.query.filter_by(crsid=current_crsid).first()
+    current_user_socs = current_user.societies
+
+    # Check the society exists
+    society = Society.query.filter_by(uid=short_name).first()
+    society_admins = society.admins
+    if not society:
+        abort(404)
+    # Check that the currently authenticated user is allowed to administer this society.
+    if not society in current_user_socs:
+        abort(403)
+
     if request.method == "POST":
 
-        name =                 request.form["name"]
-        new_short_name =       request.form["short_name"]
-        website =              request.form["website"]
-        description =          request.form["description"]
-        welcome_text =         request.form["welcome_text"]
-        logo =                 request.form["logo"]
-        banner_text =          request.form["banner_text"]
-        banner_color =         request.form["banner_color"]
-        mute_on_start =        request.form["mute_on_start"]
-        disable_private_chat = request.form["disable_private_chat"]
-
-        errors = []
-
-        if not name:
-            errors.append("A society name is required.")
-        if not new_short_name:
-            errors.append("A short society name is required.")
-
-        # TODO: confirm new_short_name is unique
-
-        if not errors:
-            pass
-            # TODO: write to the database
+        values = {}
+
+        for key in ("first_name", "surname", "email_address", "crsid"):
+            values[key] = request.form.get(key, "").strip()
+
+        errors = {}
+
+        # If the user being added is already registered, then simply get that
+        # add this society to that user's list of societies
+
+        # TODO: identifying users should be CRSid-based, changed to email address for testing purposes
+        existing_user = User.query.filter_by(email=values["email_address"]).first()
+        if existing_user:
+            # Check the existing user isn't already an admin
+            if not society in existing_user.societies:
+                pass
+            else:
+                errors["crsid"] = "This user is already an administrator."
```

Comment:
I think this section could benefit from rework as follows:

  * The first 'if not x: pass; else: do_y().` block is a bit confusing.  `if x: do_y()` would make more sense.
  * I think if the user is already an administrator there's no point validating the other bits of input (or indeed displaying validation errors for them), so we can skip forward to rendering the template.  I'd usually be annoyed by a form that doesn't show validation errors across all of its input, but perhaps we can make an exception here?

Having said all that, perhaps what we really want is a two-page flow: the first just asks for the CRSid.  Then it pops open a generic 'edit' page, using existing user data if we have it or otherwise leaving the form blank.  Appropriate UI prompts can make it clear what's happening e.g. "This person is already an admin, here are their details.  Edit them as you wish then save your changes".